File: Data Structures & Algorithms/two-integer-sum/submission-2.py
# Hash Map (Two Pass)
# Time complexity : O(n)
# Space complexity : O(n)
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        indices = {}
        for i , n in enumerate(nums):
            indices[n] = i
        for i , n in enumerate(nums):
            diff = target - n 
            if diff in indices and indices[diff] != i:
                return [i , indices[diff]]
        return []

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.cpp
// mục đích là thu hẹp phạm vi để tìm điểm đặc biệt / tìm ranh giới
class Solution {
public:
    int findMin(vector<int> &nums) {
        int l = 0 , r = nums.size() - 1;
        while(l < r)
        {
            int m = l + (r - l) / 2;
            if(nums[m] < nums[r])
            {
                // khả năng chính nums[m] chính là số nhỏ 
                r = m ;
            }
            else
            {
                l = m + 1;
            }
        }
        return nums[l];
    }
};

